tests/tests/write_texture.rs

```diff
@@ -90,107 +122,119 @@ fn write_texture_subset_2d() {
-                aspect: wgpu::TextureAspect::All,
-            },
-            wgpu::ImageCopyBuffer {
-                buffer: &read_buffer,
-                layout: wgpu::ImageDataLayout {
-                    offset: 0,
-                    bytes_per_row: Some(size),
-                    rows_per_image: Some(size),
-                },
-            },
-            wgpu::Extent3d {
-                width: size,
-                height: size,
-                depth_or_array_layers: depth,
+#[test]
+#[wasm_bindgen_test]
+fn write_texture_subset_2d_mips() {
+    let format = wgpu::TextureFormat::R8Uint;
+    let mips_count = 3;
+    let tex_size = wgpu::Extent3d {
+        width: 2048,
+        height: 2048,
+        depth_or_array_layers: 1,
+    };
+    for mip_level in 0..mips_count {
+        let mip_w = tex_size.width / (1 << mip_level);
+        let mip_h = tex_size.height / (1 << mip_level);
+        let bytes_per_row = mip_w;
+        let rows_per_image = mip_h;
+
+        let mip_extent = wgpu::Extent3d {
+            width: mip_w,
+            height: 2,
+            depth_or_array_layers: tex_size.depth_or_array_layers,
+        };
+
+        let total_bytes_in_copy =
+            total_bytes_in_copy(format, bytes_per_row, rows_per_image, mip_extent);
+
+        test(
+            format,
+            LayoutDesc {
+                bytes_per_row,
+                rows_per_image,
+                total_size_in_bytes: total_bytes_in_copy,
             },
+            mip_level,
+            mips_count,
+            tex_size,
+            mip_extent,
+            mip_extent,
         );
+    }
+}
 
-        ctx.queue.submit(Some(encoder.finish()));
-
-        let slice = read_buffer.slice(..);
-        slice.map_async(wgpu::MapMode::Read, |_| ());
-        ctx.device.poll(wgpu::Maintain::Wait);
-        let data: Vec<u8> = slice.get_mapped_range().to_vec();
-
-        for byte in &data[..((size * size) as usize * 2)] {
-            assert_eq!(*byte, 1);
-        }
-        for byte in &data[((size * size) as usize * 2)..] {
-            assert_eq!(*byte, 0);
-        }
-    });
+#[test]
+#[wasm_bindgen_test]
+fn write_texture_subset_3d() {
+    let format = wgpu::TextureFormat::R8Uint;
+    let mips_count = 1;
+    let tex_size = wgpu::Extent3d {
+        width: 256,
+        height: 256,
+        depth_or_array_layers: 4,
+    };
+    let copy_size = wgpu::Extent3d {
+        width: 256,
+        height: 256,
+        depth_or_array_layers: 2,
+    };
+
+    let bytes_per_row = tex_size.width;
+    let rows_per_image = tex_size.height;
+
+    let total_bytes_in_copy = total_bytes_in_copy(format, bytes_per_row, rows_per_image, copy_size);
+
+    test(
+        format,
+        LayoutDesc {
+            bytes_per_row,
+            rows_per_image,
+            total_size_in_bytes: total_bytes_in_copy,
+        },
+        0,
+        mips_count,
+        tex_size,
+        copy_size,
+        copy_size,
```

Comment:
```suggestion
        tex_size,
```

We should copy back the whole texture.